Clean up VASP powerups, correct params in MP flows, test for `_set_u_params`, test for `_set_kspacing`, fix `_get_incar` method of `VaspInputSet`

esoteric-ephemera opened the pull request:
Closes #465.

## Summary

A few key changes:
- Added new function, `update_vasp_input_generators` to `vasp.powerups` to remove repeated code. All subsequent functions, such as `update_user_incar_settings` now rely on this base function
- Fixed broken logic in `VaspInputGenerator._get_incar`: in `_get_incar`, `_set_kspacing` was called with `previous_incar is None` which is never true, because `previous_incar = previous_incar or {}` at the top of `_get_incar`. More, this would imply that no previous incar was found, which means `from_prev` in `_set_kspacing` was its inverse. Should be `previous_incar != {}` to indicate the presence of a previous calc
- After some discussion with @Andrew-S-Rosen , we noticed that `vasp.sets.base._get_kspacing` should return `min(kspacing, 0.44)`, to ensure that if kspacing <= 0.22, it would not be increased to 0.44
- Also from discussion with @Andrew-S-Rosen , `vasp.sets.base._set_kspacing` should default to Gaussian smearing (ISMEAR = 0) rather than tetrahedron (ISMEAR = -5) for a gapped material. While the forces should be consistent between these smearing methods for an insulator, my benchmarking has shown they can differ for semi-metals / narrow gapped systems. Gaussian is probably a safer choice overall
- Added a unit test `test_set_u_params` to `tests.vasp.test_sets` to close issue [#465](https://github.com/materialsproject/atomate2/issues/465)
- Added a unit test `test_set_kspacing` to `tests.vasp.test_sets`

janosh commented:
@esoteric-ephemera @Andrew-S-Rosen Found different `kspacing` logic in `pymatgen` `MPScanRelaxSet` and [changed it](https://github.com/materialsproject/pymatgen/pull/3396/files) to the new one in this PR. I also added better test coverage. You may want to port those tests over to `atomate2`:

```py
for bandgap, kspacing in ((10, 0.44), (3, 0.4136617), (1.1, 0.3064757), (0.5, 0.2832948), (0, 0.22)):
    incar = MPScanRelaxSet(struct, bandgap=bandgap).incar
    assert incar["KSPACING"] == approx(kspacing, abs=1e-5)
    assert incar["ISMEAR"] == -5 if bandgap > 1e-4 else 2
    assert incar["SIGMA"] == 0.05 if bandgap > 1e-4 else 0.2
```

esoteric-ephemera commented:
Just added a test for `_set_kspacing` in `tests/vasp/test_sets.py`. There was some broken logic in `VaspInputGenerator._get_incar` that I fixed as well.

utf commented:
Thanks for this. Just one comment.

> Also from discussion with @Andrew-S-Rosen , vasp.sets.base._set_kspacing should default to Gaussian smearing (ISMEAR = 0) rather than tetrahedron (ISMEAR = -5) for a gapped material. While the forces should be consistent between these smearing methods for an insulator, my benchmarking has shown they can differ for semi-metals / narrow gapped systems. Gaussian is probably a safer choice overall

I think you have this backwards. ISMEAR = -5 is very important for narrow-gapped systems. If we know the system has a gap, then we should always use ISMEAR = -5.

Andrew-S-Rosen commented on src/atomate2/vasp/sets/base.py:
```diff
@@ -1125,7 +1125,7 @@ def _set_kspacing(
             else:
                 incar["SIGMA"] = user_incar_settings.get("SIGMA", 0.05)
-                incar["ISMEAR"] = user_incar_settings.get("ISMEAR", -5)
+                incar["ISMEAR"] = user_incar_settings.get("ISMEAR", 0)
```
I will let @esoteric-ephemera comment on the ISMEAR setting, but independent of that comment, if we do stick with `ISMEAR = -5`, the `SIGMA` should be removed.

Andrew-S-Rosen replied:
Oh that's subtle!

esoteric-ephemera replied:
> I think you have this backwards. ISMEAR = -5 is very important for narrow-gapped systems. If we know the system has a gap, then we should always use ISMEAR = -5.

That's true for energies but not for forces. Looking at about 370 structures from ICSD that aren't in MP, the deviations between the forces using ISMEAR = -5 and 0 persists for even narrow-gapped materials, see below.

I guess my question is if the `_set_kspacing` function is supposed to be MP-specific or agnostic. If it's agnostic, then I'll just update ISMEAR / SIGMA for the MP flows and revert this back to -5.

![PBE_ISMEAR_forces_bandgap](https://github.com/materialsproject/atomate2/assets/33381112/a4d83647-fb13-403d-aa81-515fd719622a)

utf replied:
I'm not sure what you're plotting here? How are you evaluating whether the forces are better with ISMEAR = -5 or 0?

esoteric-ephemera replied:
Plotting consistency of the smearing methods since we don't have a reference calculation to compare with. The vertical axis is $\sum_{ij} | F_j(R_i,\mathrm{ISMEAR=-5}) -  F_j(R_i,\mathrm{ISMEAR=0})|$, where $R_i$ are the ionic positions, and $j=x, y, z$. Each structure then gets one data point.

You're right that I can't say that one is better except for true metals, where we know that the forces from the tetrahedron method aren't variational. I'm intuiting from this that there are similar errors for very narrow gap insulators and semi-metals.

For systems with very small gaps, the partial occupancies from ISMEAR = 0 are problematic in that they're unphysical, but they should help in accelerating SCF convergence. You can see for the plot below (same set of structures) that the bandgap for very small gap materials is sometimes *larger* with Gaussian smearing than tetrahedron smearing. Sometimes it's less as expected (probably from the partial occupancies).

![PBE_ISMEAR_bandgap_comp](https://github.com/materialsproject/atomate2/assets/33381112/8ca12261-92ad-4b29-9a29-c66f8ab01a27)

utf replied:
Ok, thanks for clarifying. These plots are nice to see. Just to respond to your points:

> You're right that I can't say that one is better except for true metals, where we know that the forces from the tetrahedron method aren't variational. I'm intuiting from this that there are similar errors for very narrow gap insulators and semi-metals.

Semi-metals have a band that crosses the Fermi level, and hence you should use ISMEAR = 0 as you're expecting partial occupancies. However, very narrow band gap insulators don't have a band that crosses the Fermi level, hence no partial occupancies. Here using ISMEAR = -5 will give you the most accurate results.

> For systems with very small gaps, the partial occupancies from ISMEAR = 0 are problematic in that they're unphysical, but they should help in accelerating SCF convergence. 

I'm not sure I follow that they will help with accelerating SCF convergence. I believe the main thing ISMEAR = -5 helps with is k-point convergence, and avoiding spurious partial occupancies.

Just as a final note, if you have a small gap material and want to calculate dielectric properties, then ISMEAR = -5 is essential, otherwise VASP will report a metal and the calculation will fail.

So I think its best to revert this change and set ISMEAR = -5.

esoteric-ephemera replied:
SCF convergence was the wrong phrase - I meant the reduction in needed k-point density. Both methods should help with that for semi-/metals.

I'll change this back. There's a larger ongoing discussion in MP about whether to abandon ISMEAR = -5 for relaxations in insulators (the current default) in favor of ISMEAR = 0. That will require running "fully-converged" calculations for this same set.

utf replied:
> There's a larger ongoing discussion in MP about whether to abandon ISMEAR = -5 for relaxations in insulators (the current default) in favor of ISMEAR = 0.

Ok, I would strongly advise against doing that! There are basically no downsides to ISMEAR = -5 to insulators and lots of downsides to ISMEAR = 0.

> That will require running "fully-converged" calculations for this same set.

What do you mean by "fully-converged"

esoteric-ephemera replied:
Fully converged as in using a very high density of k-points, very high cutoffs, and no smearing (Gaussian + very small smearing).

Wrt downsides: I can see issues where for very small gaps, you get erroneous partial occupancies near the fermi level. Practically, the gap has to be <~ 0.15 eV for the Gaussian smearing we use (SIGMA = 0.05 by default) for virtual states to be significantly partially occupied. Are there other downsides you're thinking of?

utf replied:
Thanks for sharing that paper. Looks interesting. 

> Are there other downsides you're thinking of?

To be honest, partial occupancies is the main one. The nice thing is that if you use ISMEAR = -5 for all gapped materials then at least your calculations are then consistent between relaxation, static, and DOS.

Can I ask why you're thinking of moving away from ISMEAR = -5?

esoteric-ephemera replied:
Two major reasons: since we already have a big transition in going from atomate --> atomate2, we're auditing the parameters MP uses in its WFs (esp. GGA); and there are a few collaborations on dataset generation where this ISMEAR benchmark was requested

janosh commented:
@esoteric-ephemera Could you submit e0186dc in a separate PR?
